### PR TITLE
Add CrossQ

### PR DESCRIPTION
**Pull request by araffin**
## Description

Implementing https://openreview.net/forum?id=PczQtTsTIX
on top of #21 

Discussion in #36 

perf report: 
https://wandb.ai/openrlbenchmark/sbx/reports/CrossQ-SBX-Perf-Report--Vmlldzo3MzQxOTAw

## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [ ] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [ ] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [ ] I have reformatted the code using `make format` (**required**)
- [ ] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [ ] I have ensured `make pytest` and `make type` both pass. (**required**)
- [ ] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

**Comment by araffin** [comment]
@danielpalen after reading the paper, I'm wondering if you have the learning curves for `relu6`?
or is it similar to SAC - TN + tanh?

File: sbx/common/jax_layers.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+                # Note: in the original paper, after some warmup phase (batch norm phase of 5k steps)
+                # the constraints are linearly relaxed to r_max/d_max over 40k steps
+                # Here we only have a warmup phase
+                is_warmed_up = jnp.greater_equal(steps.value, self.warm_up_steps).astype(jnp.float32)
+                custom_var = is_warmed_up * affine_var + (1.0 - is_warmed_up) * custom_var
+                custom_mean = is_warmed_up * affine_mean + (1.0 - is_warmed_up) * custom_mean
+
+                ra_mean.value = self.momentum * ra_mean.value + (1.0 - self.momentum) * mean
+                ra_var.value = self.momentum * ra_var.value + (1.0 - self.momentum) * var
+                steps.value += 1
```
Comment (jan1854) [inline sbx/common/jax_layers.py:190]: The variable names are a bit confusing here. `mean` and `var` are sometimes used for the batch mean and variance, sometimes for the running mean / variance. The `if not self.is_initializing()` block uses mostly `mean` / `var`, but at the end of the block `custom_mean` / `custom_var` is used, even though the variables should hold the same values.
I suggest `batch_mean` / `batch_var` for the batch statistics and to use `custom_mean` / `custom_var` only for the result of the block. I also added an explicit `if self.is_initializing()` branch to make explicit that the batch statistics are used here during initialization.

```suggestion
        if use_running_average:
            custom_mean = ra_mean.value
            custom_var = ra_var.value
        else:
            batch_mean, batch_var = _compute_stats(
                x,
                reduction_axes,
                dtype=self.dtype,
                axis_name=self.axis_name if not self.is_initializing() else None,
                axis_index_groups=self.axis_index_groups,
                # use_fast_variance=self.use_fast_variance,
            )
            if self.is_initializing():
                custom_mean = batch_mean
                custom_var = batch_var
            else:
                std = jnp.sqrt(batch_var + self.epsilon)
                ra_std = jnp.sqrt(ra_var.value + self.epsilon)
                # scale
                r = jax.lax.stop_gradient(std / ra_std)
                r = jnp.clip(r, 1 / r_max.value, r_max.value)
                # bias
                d = jax.lax.stop_gradient((batch_mean - ra_mean.value) / ra_std)
                d = jnp.clip(d, -d_max.value, d_max.value)

                # BatchNorm normalization, using minibatch stats and running average stats
                # Because we use _normalize, this is equivalent to
                # ((x - x_mean) / sigma) * r + d = ((x - x_mean) * r + d * sigma) / sigma
                # where sigma = sqrt(var)
                affine_mean = batch_mean - d * jnp.sqrt(batch_var) / r
                affine_var = batch_var / (r**2)

                # Note: in the original paper, after some warmup phase (batch norm phase of 5k steps)
                # the constraints are linearly relaxed to r_max/d_max over 40k steps
                # Here we only have a warmup phase
                is_warmed_up = jnp.greater_equal(steps.value, self.warm_up_steps).astype(jnp.float32)
                custom_mean = is_warmed_up * affine_mean + (1.0 - is_warmed_up) * batch_mean
                custom_var = is_warmed_up * affine_var + (1.0 - is_warmed_up) * batch_var

                ra_mean.value = self.momentum * ra_mean.value + (1.0 - self.momentum) * batch_mean
                ra_var.value = self.momentum * ra_var.value + (1.0 - self.momentum) * batch_var
                steps.value += 1
```

File: sbx/crossq/policies.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+        n_critics: int = 2,
+        share_features_extractor: bool = False,
+    ):
+        if optimizer_kwargs is None:
+            # Note: the default value for b1 is 0.9 in Adam.
+            # b1=0.5 is used in the original CrossQ implementation and is found
+            # but shows only little overall improvement.
+            optimizer_kwargs = {}
+            if optimizer_class in [optax.adam, optax.adamw]:
+                optimizer_kwargs["b1"] = 0.5
```
Comment (jan1854) [inline sbx/crossq/policies.py:161]: Here, the default value of `b1` is set to `0.5` only if no other arguments (or even an empty dict) are passed to the optimizer. It would be cleaner to set the default value to `0.5` regardless of the other optimizer parameters. 

```suggestion
        if optimizer_kwargs is None:
            optimizer_kwargs = {}
        if optimizer_class in [optax.adam, optax.adamw] and "b1" not in optimizer_kwargs:
            # Note: the default value for b1 is 0.9 in Adam.
            # b1=0.5 is used in the original CrossQ implementation but shows only little overall improvement.
            optimizer_kwargs["b1"] = 0.5
```

File: sbx/common/jax_layers.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+import jax.numpy as jnp
+from flax.linen.module import Module, compact, merge_param
+from flax.linen.normalization import _canonicalize_axes, _compute_stats, _normalize
+from jax.nn import initializers
+
+PRNGKey = Any
+Array = Any
+Shape = Tuple[int, ...]
+Dtype = Any  # this could be a real type?
+Axes = Union[int, Sequence[int]]
```
Comment (jan1854) [inline sbx/common/jax_layers.py:13]: Flax v0.8.1 introduced [flax.typing](https://github.com/google/flax/pull/3624), which we could use here for more descriptive type hints, similar to the current version of [flax.linen.normalization](https://flax.readthedocs.io/en/latest/_modules/flax/linen/normalization.html). However, we should probably wait a bit here since this would require a relatively recent flax version.

File: sbx/crossq/policies.py (hunk 2 above)
Comment (araffin) [reply sbx/crossq/policies.py:161]: I would keep it as is to be consistent with what is done in the rest of SB3.

File: sbx/common/jax_layers.py (hunk 1 above)
Comment (araffin) [reply sbx/common/jax_layers.py:190]: I totally agree on that point, thanks for the suggestion =)

**Comment by araffin** [comment]
@danielpalen Some early results of DroQ + CrossQ (only 2 random seeds on 3 pybullet envs, need more runs): https://wandb.ai/openrlbenchmark/sbx/reports/DroQ-CrossQ-SBX-Perf-Report--Vmlldzo3MzcxNDUy

I also quickly checked the warmup steps and could see an impact on `AntBulletEnv-v0` only when it was too small.

**Comment by danielpalen** [comment]
> @danielpalen after reading the paper, I'm wondering if you have the learning curves for `relu6`? or is it similar to SAC - TN + tanh?

I quickly checked and it looked pretty similar.

**Comment by araffin** [comment]
> I have also played around with REDQ/DroQ + CrossQ on MuJoCo but from what I remember, the results were not really consistent, sometimes better, sometimes worse.

So far, it always improved the results in my case (need more seeds to confirm, I have tried on different pybullet and mujoco envs), or at least to quickly get "good enough" solution (using up to 2x less samples than CrossQ).

One last point in case you missed it (because from https://github.com/araffin/sbx/pull/36#issuecomment-2027392759):
@danielpalen  would you be interested in providing a PyTorch implementation for SB3 contrib? (https://github.com/Stable-Baselines-Team/stable-baselines3-contrib)

**Comment by danielpalen** [comment]
> One last point in case you missed it (because from [#36 (comment)](https://github.com/araffin/sbx/pull/36#issuecomment-2027392759)): @danielpalen would you be interested in providing a PyTorch implementation for SB3 contrib? (https://github.com/Stable-Baselines-Team/stable-baselines3-contrib)

Yes, absolutely :) I put it on my todo. But I think I won't be able to get on that right away at the moment.
